### ISSUE TITLE
Enhancement Vector Effect Non-Scaling-Stroke

### ISSUE THREAD

**Issue by gabelerner** [canvg/canvg#108]
_From [kerry...@gmail.com](https://code.google.com/u/117228497632754547883/) on September 16, 2011 20:13:43_

I know this is out of the svg 1.1 spec, but I think this would be handy to have in canvg. http://www.w3.org/TR/SVGTiny12/painting.html (see section 11.5)

I started on a partial implementation to support this, and have attached the file to this issue. non-scaling-strokes should work on lines, polylines, circles, ellipses, rectangles, and paths (except for the arc command). This works with non-uniform scaling.

I believe chrome and opera already support this, so you can compare the sample svg files against them. 

Thanks,
Kerry

**Attachment:** [canvg_with_non_scaling_strokes.js cubicbezier.svg QuadBezier.svg polygon.svg rect.svg line.svg polyline.svg circles.svg ellipse.svg](http://code.google.com/p/canvg/issues/detail?id=108)

_Original issue: http://code.google.com/p/canvg/issues/detail?id=108_

**Comment by RalphMack** [comment]
We tried applying and using this patch. It does fine for simple stuff but scaling and rotation of dashed lines end up with strange line widths. More work is needed. We looked into it.The trick used in the patch is to withhold scaling from the figure and apply it instead to the vertices. The code picks up the scale factors used in the transform, saving them as a scaleX and scaleY, applies their inverses to de-scale the figure, and then scales the vertices. Unfortunately, when you do that, the effects of scaling on translation and rotation aren't fully taken into account. At a minimum, the non-transitivity of subsequent transforms is overlooked. The cut points for change are probably right, but the solution will need to be more complex. There's an exercise in linear algebra waiting to happen here. (Oh - one other thing - it broke elliptical (A) curves in paths so we changed our own stuff to use cubic beziers for what we needed - that part is probably an easier fix.)
